Product Creation AI: Use prompt to generate product

selanthiraiyan opened the pull request:
Closes: #10767 

## Description

Adds a new AI prompt to ask AI to generate a product as a JSON and parses that JSON response into a `Product`

#### Changes
- New AI feature added to keep track of AI usage and costs of the product creation feature. 
- Added a new prompt to generate product details.
- Added a new mapper to map the JSON response from Jetpack AI into a `Product`.
- Mock JSON files for testing.
- Made `ProductTag` copiable for unit testing purposes. 
- Unit tests for remote and mapper.

Sorry about the 500+ lines PR. It is due to the mock files and unit tests. 🙏 

## Testing instructions
CI passing is sufficient. We will use this API in a future PR.

## Screenshots
NA

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Add mapper to parse AI JSON response.

selanthiraiyan committed:
Make ProductTag copiaable.

selanthiraiyan committed:
New feature case for product creation.

selanthiraiyan committed:
Use AI prompt to generate product.

selanthiraiyan committed:
Add generate product mock JSON responses.

selanthiraiyan committed:
Unit test generateProduct GenerativeContentRemote method.

selanthiraiyan committed:
Add unit tests for `AIProductMapperTests`.

selanthiraiyan committed:
Merge branch 'trunk' into feat/10767-ai-prompt

itsmeichigo commented on Networking/Networking/Mapper/AIProductMapper.swift:
```diff
@@ -0,0 +1,223 @@
+                  dateOnSaleStart: nil,
+                  dateOnSaleEnd: nil,
+                  productTypeKey: ProductType.simple.rawValue,
+                  statusKey: ProductStatus.published.rawValue,
```
👀 Do you think we should keep the status as draft so that we don't have to update it later when saving it as draft on the UI layer?

selanthiraiyan replied:
I agree. Done in https://github.com/woocommerce/woocommerce-ios/pull/10770/commits/b651a6f817eee96c3d2e93066d743413106e9a84

itsmeichigo commented on Networking/Networking/Remote/GenerativeContentRemote.swift:
````diff
@@ -199,80 +200,71 @@ private extension GenerativeContentRemote {
+            }
+
+            return
+                ", tags:Given the list of available tags ```\(tags.map { $0.name }.joined(separator: ", "))```, " +
````
❓ Any reason why we don't have a space after the colon? Is that to make AI work?

selanthiraiyan replied:
Added spaces in fc4f2edaa76125215ffcaf55690743c410aa8455

itsmeichigo commented on Networking/Networking/Remote/GenerativeContentRemote.swift:
````diff
@@ -199,80 +200,71 @@ private extension GenerativeContentRemote {
+            return
+                ", tags:Given the list of available tags ```\(tags.map { $0.name }.joined(separator: ", "))```, " +
+                "suggest an array of the best matching tags for this product, if no matches are found return an empty array, " +
+                "don’t suggest any value other than the available ones"
````
👀 Please add a period mark at the end of the sentence to make the prompt readable.

selanthiraiyan replied:
Done in 1e61884e951eb7fa2832e4e2ecb36c89fa9777a7

itsmeichigo commented on Networking/NetworkingTests/Remote/GenerativeContentRemoteTests.swift:
```diff
@@ -226,6 +226,287 @@ final class GenerativeContentRemoteTests: XCTestCase {
+                                                        tags: [ProductTag.fake(), ProductTag.fake()])
+
+        // Then
+        XCTAssertNotNil(product)
```
Nit: Can we check and compare product details too?

selanthiraiyan replied:
Done in 0e1402521d415ff4387a0711269114856f92ae82

itsmeichigo reviewed (approved):
Thanks for the thorough tests! I left some comments below but approving to unblock development.

selanthiraiyan committed:
Change new AI product status to draft.

selanthiraiyan committed:
Add space after colon.

selanthiraiyan committed:
Add period marks at the end of sentence.

selanthiraiyan committed:
Unit test all product properties.

selanthiraiyan commented:
Thanks for the review, Huong! I have addressed your comments. I am enabling auto-merge now. Please feel free to add comments if needed. I will address those in a future PR.

selanthiraiyan committed:
Merge branch 'trunk' into feat/10767-ai-prompt
